### PR TITLE
feat(core): Allow Menus to return values

### LINKED ISSUE #7370: Allow choosing more actions via external menus
Currently, `ui.layouts.menu.Menu` supports only showing information and cancelling the flow.

It should support returning more actions, which can be useful for handling `confirm_blob_intro` flow, which currently can return `CONFIRMED`, `CANCELLED` or `INFO` - but we'd like it also to return `HASH` (for #6680).

Image

Image

### PR DESCRIPTION
**Pull request by vojczejk**
Allow menu leaves to return typed values in preparation for modularisation of the firmware and UI.

Reimagines the menu as a proper tree. The menu is now able to, based on user interaction, return an arbitrary value and infrormation about from which submenu/menu leaf it comes from.

Menu[R] is the internal node, MenuLeaf[R] (renamed from Details) is the leaf, MenuNode[R] is a type alias for either. R is the type of the value the menu tree returns.

A leaf's interact() returning None still just resumes the tree (pops one level), as before (usually pressing the back button). 

Returning a value makes show_menu() unwind the whole tree and report a MenuResult pairing the value with the leaf that produced it, so callers with multiple value-returning leaves can match on leaf identity first, then value.

interact_with_menu() now returns the MenuResult instead of re-showing the main layout; it keeps a temporary overload for the ~25 existing call sites that only use original menus and pass the result straight through. This is marked explicitly as compatibility shim and is due for removal during further rework.

No behavior change: every existing menu is still built from information-only leaves that discard their layout's result.

Update: Reworked the cancellation to be more generic, created the notion of an intent for each menu item. This shall translate to the level of the urgency conveyed by the menu item. Currently the only intents are Standard and Danger, with standard being the usual white menu items and Danger being used for the red cancellation menu items. 

# Notes 4 QA

The PR is done such that there should be 0 functional user experienceable change. 

## Scope by model

- T3T1, T3W1 — all menus, main targets.
- T2B1 / T3B1 — never drew a cancel entry, does not style by intent. Smoke test only.
- T2T1 — no menus in the Python layer. Not affected.

## Affected paths

- Sign message, T3T1 — `trezorctl btc sign-message -n "m/44h/0h/0h/0/0" "hello"`
- Send transaction, T3T1 + T3W1 (Suite) — menu on both address and amount screens
- Ethereum transaction and staking, T3T1 + T3W1
- Ethereum EIP-7702 authorise / revoke, T3T1 + T3W1
- Ethereum vault claim / transaction, T3T1 + T3W1
- Solana staking, T3W1 + T2B1/T3B1
- Payment request and swap, T3T1 + T3W1 (Suite)
- Trade confirmation, T3T1 + T3W1
- Tron claim, T3T1 + T3W1

## Edge cases - stuff that could break (or did break during development)

- Menu at capacity — payment request with provider address + 4 refunds is exactly
  6 entries. Renders and scrolls. Limit went 5 → 6 only because cancel is now a real menu entry, not a special case. The number of information entries is unchanged.
- Dismissing an information entry with the header close button returns to
  the menu. On `main` this could escape and read as cancelling the whole flow.
- Cancel entry with confirmation, output screen — decline returns to the menu,
  accept aborts signing with a cancellation on the host.
- Cancel entry position — T3T1 first, T3W1 last, same as `main`.
- Cancel entry label, T3W1 output screen — "Cancel sign", not "Cancel".
- Closing the menu itself returns to the confirmation screen, never aborts.

## Testing done

- Unit tests as found in branch vojczejk/ui-menu-ut
- Builds: firmware and bootloader for T2T1, T2B1, T3T1, T3W1.
- Not run locally: Already present UI tests for all models.

### CODE REVIEW

**Comment by vojczejk** [comment]
A branch demonstrating use of the Menu is now available vojczejk/ui-menu-generalisation-demo

To invoke the menu, invoke any interaction that uses it, e.g. 
`trezorctl crypto get-entropy 32` and click on the menu.

File: core/src/trezor/ui/layouts/menu.py [hunk 1]
```diff
@@ -33,45 +41,87 @@ def __init__(
...
+    root: Menu[R],
     raise_on_cancel: ExceptionType = ActionCancelled,
-) -> None:
+) -> MenuResult[R] | None:
+    """Walk the menu tree until a leaf produces a value, or the user leaves the root.
+
+    Returns the leaf and the value it produced, so that the caller can tell the
+    leaves apart, or `None` if the tree was left without producing a value.
+    """
     menu_path = []
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py]: nit: let's document its type (mainly for readability)

```suggestion
    menu_path: list[int] = []
```

File: core/src/trezor/ui/layouts/caesar/__init__.py [hunk 2]
```diff
@@ -2591,9 +2596,11 @@ async def confirm_firmware_update(description: str, fingerprint: str) -> None:
         )
 
 
-def create_details(name: str, value: Sequence[StrPropertyType] | str) -> Details:
-    from trezor.ui.layouts.menu import Details
+def create_info_menu_leaf(
+    name: str, value: Sequence[StrPropertyType] | str
+) -> MenuLeaf:
```
Comment (romanz) [inline core/src/trezor/ui/layouts/caesar/__init__.py]: nit: prefer an explicit typing (here and in other layouts), otherwise IIUC it will be treater as `MenuLead[Unknown]`
```suggestion
) -> MenuLeaf[None]:
```


https://github.com/user-attachments/assets/340797b0-b4d6-4833-8566-08d93f027bef

File: core/src/trezor/ui/layouts/menu.py [hunk 3]
```diff
@@ -33,45 +41,87 @@ def __init__(
 
     @classmethod
     def root(
-        cls, children: Iterable["Details"] = (), cancel: "str | Cancel | None" = None
+        cls,
+        children: "Iterable[MenuNode[R]]" = (),
+        cancel: "str | Cancel | None" = None,
     ) -> Self:
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py]: It seems that VSCode think that  `Menu[Unknown]` is returned here... can we fix it?


https://github.com/user-attachments/assets/acb089cf-1fe3-4634-8f64-96e27731d769

**Review by romanz — COMMENTED** [review]
A few typing-related comments:

File: core/src/trezor/ui/layouts/menu.py (hunk 1 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py]: Done. Now menu_path: list[int] = [] at line 131.

File: core/src/trezor/ui/layouts/caesar/__init__.py (hunk 2 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/caesar/__init__.py]: Done in all three: caesar:2587, delizia:2494, eckhart:2623.

File: core/src/trezor/ui/layouts/menu.py (hunk 3 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py]: The Menu.root() classmethod this was on is gone — Menu is constructed directly now, so the Self return that was dropping the parameter no longer exists.

Checked the replacement with reveal_type under core/pyrightconfig.json:
```
Menu()                              -> Menu[Unknown]
a: Menu = Menu()                    -> Menu[None]
Menu([])                            -> Menu[None]
Menu([cancel_leaf("Cancel")])       -> Menu[Never]
Menu([info, cancel_leaf("Cancel")]) -> Menu[None]
Menu([cancel_leaf("Cancel"), info]) -> Menu[None]
```

**Comment by vojczejk** [comment]
The branch vojczejk/ui-menu-generalisation-demo contains up-to-date demo hidden in the ethereum sign flow.

to reach it, you can use

`trezorctl ethereum sign-tx -c 1 -n "m/44h/60h/0h/0/0" -g 21000 -G 20000000000 -i 0 x8eA7a3fccC211ED48b763b4164884DDbcF3b0A98 1000000000`

or other ethereum sign-tx

File: core/embed/rust/src/ui/api/firmware_micropython.rs [hunk 4]
```diff
@@ -1868,11 +1871,11 @@ pub static mp_module_trezorui_api: Module = obj_module! {
 
     /// def select_menu(
     ///     *,
-    ///     items: Iterable[str],
+    ///     items: Iterable[tuple[str, int]],
```
Comment (romanz) [inline core/embed/rust/src/ui/api/firmware_micropython.rs:1875]: Can we make the `int` into an `IntEnum` (for type checking)?

File: core/embed/rust/src/ui/layout_bolt/component/select_menu.rs [hunk 5]
```diff
@@ -6,63 +6,55 @@ use crate::strutil::TString;
...
     choice_buttons: Vec<Button, MAX_MENU_ITEMS>,
-    cancel_button: Option<Button>,
     close_button: Button,
 }
 
 impl SelectMenu {
     pub fn new(
-        items: Vec<TString<'static>, MAX_MENU_ITEMS>,
-        cancel: Option<TString<'static>>,
+        items: Vec<(TString<'static>, MenuItemIntent), MAX_MENU_ITEMS>,
```
Comment (romanz) [inline core/embed/rust/src/ui/layout_bolt/component/select_menu.rs]: nit: encapsulate `(TString<'static>, MenuItemIntent)` in a `MenuItem` struct?

File: core/embed/rust/src/ui/layout_delizia/ui_firmware.rs [hunk 6]
```diff
@@ -718,34 +718,27 @@ impl FirmwareUI for UIDelizia {
...
         let mut menu_items = VerticalMenuItems::new();
-        if let Some(text) = cancel {
-            unwrap!(menu_items.push(VerticalMenuItem::Cancel(text)));
-            current += 1;
-        }
-        for text in items {
-            unwrap!(menu_items.push(VerticalMenuItem::Item(text)));
+        for (text, intent) in items {
+            unwrap!(menu_items.push(match intent {
+                // TODO: mapping `Danger` onto the `Cancel` variant is temporary -
```
Comment (romanz) [inline core/embed/rust/src/ui/layout_delizia/ui_firmware.rs:728]: Let's rename it in a separate commit/PR?
(here and below)

File: core/embed/rust/src/ui/ui_firmware.rs [hunk 7]
```diff
@@ -18,6 +25,48 @@ pub const MAX_MENU_ITEMS: usize = 5;
 
 pub const MAX_PAIRED_DEVICES: usize = 8; // Maximum number of paired devices in the device menu
 
+/// What a menu entry means, which each model renders in its own way.
+#[repr(u8)]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
```
Comment (romanz) [inline core/embed/rust/src/ui/ui_firmware.rs]: I guess we don't need `Debug` in production builds...

If we need it for tests, we can use:
https://github.com/trezor/trezor-firmware/blob/809dfef2b48a6cbcb9186471899b53345576a065/core/embed/crypto/src/lib.rs#L18

File: core/src/trezor/ui/layouts/caesar/__init__.py [hunk 8]
```diff
@@ -1517,12 +1521,12 @@ async def confirm_ethereum_eip7702_auth(
...
             children = [
-                create_details(TR.address_details__account_info, account_info),
-                create_details(TR.buttons__more_info, more_info),
+                create_info_menu_leaf(TR.address_details__account_info, account_info),
+                create_info_menu_leaf(TR.buttons__more_info, more_info),
             ]
             await confirm_with_menu(
                 layout,
-                Menu.root(children, cancel=TR.buttons__cancel),
+                Menu(children + [cancel_leaf(TR.buttons__cancel)]),
```
Comment (romanz) [inline core/src/trezor/ui/layouts/caesar/__init__.py]: nit: add the `cancel_leaf` to `children` before calling `interact_with_menu`, to avoid allocating a new list.
(here and below)

File: core/src/trezor/ui/layouts/menu.py [hunk 9]
```diff
@@ -8,130 +8,230 @@
...
+    (one level up), or a value of type `R`, which is propagated out of the whole
+    tree by `show_menu()`.
+    """
 
-class Details:
-    def __init__(self, name: str, interact: Callable[[], Awaitable[T]]) -> None:
+    def __init__(
+        self,
+        name: str,
+        interact: Callable[[], Awaitable[R | None]],
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py:58]: What happens if `R` is `None`?

File: core/src/trezor/ui/layouts/menu.py [hunk 10]
```diff
@@ -8,130 +8,234 @@
...
+        *,
+        intent: int = trezorui_api.MenuItemIntent.STANDARD,
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py:60]:
```suggestion
        intent: trezorui_api.MenuItemIntent = trezorui_api.MenuItemIntent.STANDARD,
```
(here and below)

File: core/src/trezor/ui/layouts/menu.py [hunk 11]
```diff
@@ -8,130 +8,230 @@
...
-                # details' layout is de-allocated after interact() returns.
-                return await interact(obj, br_name=None, raise_on_cancel=None)
+def leaf_from_layout(
+    name: str,
+    layout_factory: Callable[[], trezorui_api.LayoutContext[R]],
+    *,
+    return_result: bool = False,
+    br_name: str | None = None,
+    br_code: ButtonRequestType = ButtonRequestType.Other,
+    raise_on_cancel: ExceptionType | None = None,
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py:76]: Is it safe to ignore cancellation by default when `return_result == True`?
(since `raise_on_cancel` defaults to `None`)

File: core/src/trezor/ui/layouts/menu.py [hunk 12]
```diff
@@ -8,130 +8,230 @@
...
+    # declare `-> UiResult` and hand the result straight back, so the union does not
+    # typecheck there. Every one of them passes an info-only menu (`create_info_menu_leaf()`
+    # leaves, which discard their layout's result), so the first overload keeps them
+    # on the old plain-`T` typing while the second serves menus that do produce a
+    # value.
+    #
+    # Drop both overloads once these call sites handle `MenuResult` themselves
+    # (25 calls in total, all of them value-less menus today):
+    #
+    #   delizia:  confirm_action, confirm_value, confirm_ethereum_vault_tx (5),
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py]: nit: those numbers may change over time, so it's better to keep those in the commit message.

File: core/embed/rust/src/ui/layout_caesar/ui_firmware.rs [hunk 13]
```diff
@@ -912,17 +913,23 @@ impl FirmwareUI for UICaesar {
     }
 
     fn select_menu(
-        items: heapless::Vec<TString<'static>, MAX_MENU_ITEMS>,
+        items: heapless::Vec<(TString<'static>, MenuItemIntent), MAX_MENU_ITEMS>,
         current: usize,
-        _cancel: Option<TString<'static>>,
     ) -> Result<impl LayoutMaybeTrace, Error> {
+        // the entry's intent is not rendered on this model
+        let labels: heapless::Vec<TString<'static>, MAX_MENU_ITEMS> =
```
Comment (romanz) [inline core/embed/rust/src/ui/layout_caesar/ui_firmware.rs:920]: nit (it seems to work without type annotations):

```suggestion
        let labels =
```

File: core/src/trezor/ui/layouts/delizia/__init__.py [hunk 14]
```diff
@@ -532,7 +532,7 @@ async def confirm_payment_request(
                 refund_account_items,
             )
         )
-    menu = Menu(menu_items + [cancel_leaf(TR.buttons__cancel_sign)])
+    menu = Menu([cancel_leaf(TR.buttons__cancel_sign)] + menu_items)
```
Comment (romanz) [inline core/src/trezor/ui/layouts/delizia/__init__.py]: Add the `cancel_leaf` to `children` before calling `interact_with_menu`, to avoid allocating a new list.
(here and elsewhere)

File: core/embed/rust/src/ui/layout_caesar/component/input_methods/simple_choice.rs [hunk 15]
```diff
@@ -9,8 +9,15 @@ use crate::ui::geometry::Rect;
...
-// as would be via `const N: usize` generics.
-const MAX_LENGTH: usize = 5;
+// as would be via `const N: usize` generics. Callers must size their `Vec`
+// with this constant rather than repeating the literal.
+pub const MAX_LENGTH: usize = 6;
+
+/// `select_menu()` renders through `SimpleChoice`, handing it a list already
+/// bounded by `MAX_MENU_ITEMS`. Keep the capacity at or above that bound so
+/// raising `MAX_MENU_ITEMS` alone cannot stop this model from building.
+#[cfg(feature = "micropython")]
```
Comment (romanz) [inline core/embed/rust/src/ui/layout_caesar/component/input_methods/simple_choice.rs:19]: nit: why limit this check to MicroPython only?

File: core/src/trezor/ui/layouts/menu.py [hunk 16]
```diff
@@ -8,130 +8,234 @@
...
+    is resumed, as with any leaf that returns `None`.
+    """
+
+    async def _interact() -> None:
+        if confirm is not None:
+            with confirm() as obj:
+                result = await interact(obj, br_name=None, raise_on_cancel=None)
+            if result is not trezorui_api.CONFIRMED:
+                return None  # declined - back to the menu
+        raise exc
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py]: Let's crash on unexpected `result` values:
```python
        if confirm is None:
            raise exc

        with confirm() as obj:
            result = await interact(obj, br_name=None, raise_on_cancel=None)
        if result is trezorui_api.CANCELLED:
            return None  # declined - back to the menu
        if result is trezorui_api.CONFIRMED:
            raise exc
        raise RuntimeError
```

File: core/src/trezor/ui/layouts/menu.py [hunk 17]
```diff
@@ -8,130 +8,234 @@
...
-                if menu.cancel:
-                    result = await menu.cancel._interact()
-                    assert result in (trezorui_api.CONFIRMED, trezorui_api.CANCELLED)
-                    if result is trezorui_api.CONFIRMED:
-                        # cancellation is confirmed - raise an exception
-                        raise raise_on_cancel
-                    # cancellation is not confirmed - back to the menu
-                    continue
-            elif isinstance(choice, int):
+            if isinstance(choice, int):
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py:179]: Crash if it's an unexpected type/value?

**Review by romanz — COMMENTED** [review]
Concept ACK - many thanks for the refactoring!
The menu-related code looks much better now :)

File: core/embed/rust/src/ui/layout_caesar/ui_firmware.rs (hunk 13 above)
Comment (vojczejk) [reply core/embed/rust/src/ui/layout_caesar/ui_firmware.rs:920]: I prefer self documenting code and explicit (and usually redundant) type annotations provide that as well as explicit type assertions.

File: core/embed/rust/src/ui/layout_bolt/component/select_menu.rs [hunk 18]
```diff
@@ -14,55 +15,47 @@ const MAX_VISIBLE_BUTTONS: usize = 3;
...
     ) -> Result<Self, Error> {
-        if items.len() + cancel.map_or(0, |_| 1) > 3 {
+        if items.len() > MAX_VISIBLE_BUTTONS {
             return Err(Error::NotImplementedError);
         }
```
[the remark this answers is not in the document]
Comment (vojczejk) [reply core/embed/rust/src/ui/layout_bolt/component/select_menu.rs:35]: Pagination is the only correct answer here and its out of scope of this ticket. This edge case is not reached and would easily be discovered in tests if it is.

The location is marked with appropriate comment.

File: core/embed/rust/src/ui/api/firmware_micropython.rs (hunk 4 above)
Comment (vojczejk) [reply core/embed/rust/src/ui/api/firmware_micropython.rs:1875]: I agree on the goal but oppose the change as you suggest unless its an codebase-unifying/wide change. There are other cases (e.g. NotificationLevel) that do it this way already. I would prefer not to introduce new inconsistency despite more local correctness.

Comment (romanz) [reply core/embed/rust/src/ui/api/firmware_micropython.rs:1875]: I agree - let's keep it for now.

File: core/embed/rust/src/ui/layout_bolt/component/select_menu.rs (hunk 5 above)
Comment (vojczejk) [reply core/embed/rust/src/ui/layout_bolt/component/select_menu.rs]: Done, named SelectMenuItem to avoid collision with MenuItem used in eckharts menus. once unified, we can rename to generic MenuItem

File: core/embed/rust/src/ui/layout_delizia/ui_firmware.rs (hunk 6 above)
Comment (vojczejk) [reply core/embed/rust/src/ui/layout_delizia/ui_firmware.rs:728]: Agreed, should I create a separate task for this? @obrusvit

File: core/embed/rust/src/ui/ui_firmware.rs (hunk 7 above)
Comment (vojczejk) [reply core/embed/rust/src/ui/ui_firmware.rs]: Done, FYI

ui/ has 11 other unconditional Debug derives:

ui/util.rs:252                                  
ui/geometry.rs:38
ui/notification.rs:33
ui/component/text/paragraphs.rs:411
ui/component/text/layout.rs:615
ui/display/image.rs:42
ui/display/toif.rs:6
ui/layout_eckhart/firmware/keyboard/label.rs:18
ui/layout_eckhart/firmware/keyboard/string.rs:254
ui/layout_eckhart/firmware/keyboard/keypad.rs:53
ui/layout_eckhart/firmware/keyboard/pin.rs:262

Separate task? Or should i just fix it all over as collateral feature? @obrusvit

File: core/src/trezor/ui/layouts/menu.py (hunk 9 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:58]: `R = None` collapses `R | None` to `None`, so such a leaf can never produce a value

That's the info-leaf case, and why `create_info_menu_leaf` returns `MenuLeaf[None]`. `show_menu` never yields a `MenuResult` for those, and the `Menu[None]` overload keeps existing callers on plain-`T`.

File: core/embed/rust/src/ui/ui_firmware.rs (hunk 7 above)
Comment (romanz) [reply core/embed/rust/src/ui/ui_firmware.rs]: Let's do it as a separate task.

File: core/embed/rust/src/ui/layout_delizia/ui_firmware.rs (hunk 6 above)
Comment (romanz) [reply core/embed/rust/src/ui/layout_delizia/ui_firmware.rs:728]: Separate PR/task :+1:

File: core/src/trezor/ui/layouts/menu.py (hunk 10 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:60]: `Type of "trezorui_api.MenuItemIntent.STANDARD" is "int"
error: Expression of type "int" cannot be assigned to parameter of type "MenuItemIntent"`

Same request as "Can we make the int into an IntEnum (for type checking)?" - goes hand-in-hand.

File: core/src/trezor/ui/layouts/delizia/__init__.py (hunk 14 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/delizia/__init__.py]: Done

File: core/src/trezor/ui/layouts/menu.py (hunk 11 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:76]: No, its not.

Added a catch case for CANCELLED with a verbose comment describing what needs to happen to fix this cleanly.

File: core/src/trezor/ui/layouts/menu.py (hunk 12 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py]: Valid, its gone.

File: core/src/trezor/ui/layouts/menu.py (hunk 16 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py]: Accepted.

File: core/src/trezor/ui/layouts/menu.py (hunk 17 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:179]: Accepted.

File: core/embed/rust/src/ui/layout_caesar/component/input_methods/simple_choice.rs (hunk 15 above)
Comment (vojczejk) [reply core/embed/rust/src/ui/layout_caesar/component/input_methods/simple_choice.rs:19]: If i understand it correctly, MAX_MENU_ITEMS lives in ui_firmware, which is itself gated:

```
// ui/mod.rs:40
#[cfg(feature = "micropython")]
pub mod ui_firmware;
Bootloader builds compile trezor_lib without micropython
```
I ran the build without and all pass except delizia which fails with:
```
error[E0433]: cannot find `ui_firmware` in `ui`
   --> rust/src/ui/layout_delizia/component/vertical_menu.rs:323:57
note: found an item that was configured out
 40 | #[cfg(feature = "micropython")]
```
Only Delizia's file is reached by a bootloader build today. I've kept the gate on all three anyway, for consistency but I can scope it tighter if it makes more sense to you?.

File: core/src/trezor/ui/layouts/menu.py (hunk 17 above)
Comment (romanz) [reply core/src/trezor/ui/layouts/menu.py:179]: I meant - let's raise an exception if `choice` is a non-integer :)

File: core/src/trezor/ui/layouts/menu.py [hunk 19]
```diff
@@ -8,130 +8,265 @@
...
                 # go one level down
                 menu_path.append(choice)
                 current_item = 0
                 continue
+            if choice is not trezorui_api.CONFIRMED:
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py]: nit: I'd write it a bit differently:
```
```

File: core/src/trezor/ui/layouts/menu.py [hunk 20]
```diff
@@ -8,130 +8,265 @@
...
+        br_name: str | None,
+        br_code: ButtonRequestType = ButtonRequestType.Other,
+        raise_on_cancel: ExceptionType = ActionCancelled,
+        *,
+        layout_type: type[Layout] = Layout,
+    ) -> T | MenuResult[R]:
+        """Either the main layout's result, or a leaf and the value it produced."""
 
 
 async def interact_with_menu(
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py:244]: Should we add a unittest for the menu functionality?

OTOH, it's also fine by me to use click/device tests when the new-style menus will be used.

File: core/src/trezor/ui/layouts/menu.py [hunk 21]
```diff
@@ -8,130 +8,265 @@
...
+        br_name: str | None,
+        br_code: ButtonRequestType = ButtonRequestType.Other,
+        raise_on_cancel: ExceptionType = ActionCancelled,
+        *,
+        layout_type: type[Layout] = Layout,
+    ) -> T:
+        """The menu cannot produce a value, so only the main layout's result."""
+
+    @overload
+    async def interact_with_menu(
```
Comment (romanz) [inline core/src/trezor/ui/layouts/menu.py:232]: Please review existing `interact_with_menu()` call-sites, to be sure the new return values will not be ignored (when new-style menus will be introduced).

e.g. we raise if `confirm_linear_flow()` gets an unexpected value - but Eckhart `confirm_tron_claim()` seems to drop the return value.

**Review by romanz — APPROVED** [review]
LGTM, many thanks!

File: core/embed/rust/src/ui/layout_caesar/component/input_methods/simple_choice.rs (hunk 15 above)
Comment (romanz) [reply core/embed/rust/src/ui/layout_caesar/component/input_methods/simple_choice.rs:19]: Thanks - let's keep it.

File: core/src/trezor/ui/layouts/menu.py (hunk 21 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:232]: Reviewed, unified tron usage, should be clean all over now.

Comment (romanz) [reply core/src/trezor/ui/layouts/menu.py:232]: Thanks!
CC: @PrisionMike

File: core/embed/rust/src/ui/layout_delizia/ui_firmware.rs (hunk 6 above)
Comment (obrusvit) [reply core/embed/rust/src/ui/layout_delizia/ui_firmware.rs:728]: @coderabbitai create an issue, assign to @vojczejk

File: core/src/trezor/ui/layouts/menu.py (hunk 20 above)
Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:244]: I have added a set of tests incl. some mocks/fakes to support tsting without the UI. there is quite some logic that can be tested without the UI so I did.

Comment (romanz) [reply core/src/trezor/ui/layouts/menu.py:244]: We can merge this PR without https://github.com/trezor/trezor-firmware/pull/7651/commits/2257510698f5c41829dd8b6bd03e3101819aedf1, let's discuss it in a separate PR :)

Comment (vojczejk) [reply core/src/trezor/ui/layouts/menu.py:244]: @coderabbitai create an issue to add UTs to this and assign to @vojczejk

File: core/embed/rust/src/ui/ui_firmware.rs (hunk 7 above)
Comment (obrusvit) [reply core/embed/rust/src/ui/ui_firmware.rs]: @coderabbitai open an issue for the debug derives as listed above.

@vojczejk in general, prefer scoped PRs to prevent feature-creep in existing opened PRs.

**Comment by Thalarion** [comment]
Test automation giving this ✅ is good enough. The change is so broad most testing will also cover this generally.
